[Misc] Add offline test for disaggregated prefill

Shaoting-Feng opened the pull request:
This PR adds an offline test for disaggregated prefill use case.

Shaoting-Feng committed:
Fix format

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

WangErXiao reviewed (approved):
good

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,80 @@
+
+    prompts = [
+        "Hello, my name is",
+        # "Hi, your name is", # To simulate transmission failure
```
Can you elaborate? How this comment simulates the failure?

Shaoting-Feng replied:
This is to trigger the partial prefill of requests in a batch. The prefill node receives two requests, while the decode node receives three requests. So the decode node will only receive the KV Cache for requests 1 and 3. The decode node will use the KV Cache of requests 1 and 3 and do prefilling on request 2. 

This example demonstrates how to use disaggregated prefill and how the decode node manages receiving only a subset of requests within a batch.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,80 @@
+
+
+def run_prefill(prefill_done):
+    os.environ["CUDA_VISIBLE_DEVICES"] = "0"
```
Please add comments to explain that we use GPU 0 for prefill and GPU1 for decode.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,80 @@
+        # "Hi, your name is", # To simulate transmission failure
+        "Tell me a very long story",
+    ]
+    sampling_params = SamplingParams(temperature=0, top_p=0.95, max_tokens=1)
```
max_token=1 seems not a very good prefill disaggregated example.

Shaoting-Feng replied:
Thanks for your comment. This example serves as a proxy API server in the offline case. As shown in the Readme of disaggregated prefill (https://github.com/vllm-project/vllm/tree/main/vllm/distributed/kv_transfer), we should set the max token of prefill node to 1. 

Also in the online case, in `benchmarks/disagg_benchmarks/disagg_prefill_proxy_server.py`, the author set `prefill_request['max_tokens'] = 1` in `handle_request` function.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,80 @@
+    llm = LLM(model="meta-llama/Meta-Llama-3.1-8B-Instruct",
+              kv_transfer_config=ktc,
+              max_model_len=2000,
+              gpu_memory_utilization=0.8)
```
Some GPUs may get OOM with this ratio. Please document the GPU you used and advise the ratio for other GPUs.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,80 @@
+    print("Prefill node is finished.")
+    prefill_done.set()
+
+    # To keep the prefill node running in case the decode node is not done
+    try:
+        while True:
+            time.sleep(1)
+    except KeyboardInterrupt:
+        print("Script stopped by user.")
```
This won't stop until user presses Carl+C right? Then this is not a good offline example...

Shaoting-Feng replied:
The prefill node will automatically terminate once the decode node completes. The code is as follows:

```python
# Terminate the prefill node when decode is finished
decode_process.join()
prefill_process.terminate()
```

comaniac reviewed (commented):
Thanks for the PR. Please note that we expect users to learn the feature from examples and ideally users can directly use or modify examples for their use cases. So please provide as many comments and explanations as possible. In addition, since it's not trivial for prefill disaggregation to be used in offline inference, could you also elaborate on the scenario?

Shaoting-Feng committed:
Add comments and explanations

Signed-off-by: Shaoting Feng <shaotingf@uchicago.edu>

Shaoting-Feng commented:
@comaniac @WangErXiao Thank you very much for your feedback. I have added additional comments and addressed your questions. An offline disaggregated prefill example use case is valuable for the disaggregated prefill roadmap, as it simplifies debugging. If there are any further concerns about the code, I would greatly appreciate your input.

Shaoting-Feng commented:
@comaniac Could I get a review again please? Btw the failing checks should be irrelevant to this PR since this PR is only an example. Thank you in advance.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,98 @@
+We will launch 2 vllm instances (GPU 0 for prefill and GPU 1 for decode),
+and then transfer the KV cache between them.
+
+Learn more about Ray Data in https://docs.ray.io/en/latest/data/data.html
```
This example has nothing to do with Ray Data?

Shaoting-Feng replied:
Yeah. I've deleted this comment.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,98 @@
+    # and 3 and do prefilling on request 2.
+    prompts = [
+        "Hello, my name is",
+        # "Hi, your name is", # To trigger partial prefill of batched requests
```
"partial prefill" is not the right term for this case.

```suggestion
        # "Hi, your name is", # The decode node will actually "prefill" this request.
```

Shaoting-Feng replied:
Thanks for your fix. I've changed this comment.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,98 @@
+    ]
+    sampling_params = SamplingParams(temperature=0, top_p=0.95, max_tokens=1)
+
+    ktc = KVTransferConfig.from_cli(
+        '{"kv_connector":"PyNcclConnector","kv_role":"kv_producer","kv_rank":0,"kv_parallel_size":2}'
+    )
```
Example what does this configure mean.

Shaoting-Feng replied:
Thanks for your comment. I've added explanations for the KV transfer config according to the `config.py`.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,98 @@
+    ktc = KVTransferConfig.from_cli(
+        '{"kv_connector":"PyNcclConnector","kv_role":"kv_producer","kv_rank":0,"kv_parallel_size":2}'
+    )
+    # Example: Set GPU memory utilization to 0.8 for an A6000 GPU with 40GB
+    # memory. Reduce the value if your GPU has less memory.
```
```suggestion
    # Set GPU memory utilization to 0.8 for an A6000 GPU with 40GB
    # memory. You may need to adjust the value to fit your GPU.
```

Shaoting-Feng replied:
Thanks for your comment. I've changed this comment.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,98 @@
+    print("Prefill node is finished.")
+    prefill_done.set()
+
+    # To keep the prefill node running in case the decode node is not done
```
```suggestion
    # To keep the prefill node running in case the decode node is not done; otherwise ...
```

Shaoting-Feng replied:
Thanks for your comment. I've completed this comment.

comaniac commented on examples/offline_inference/disaggregated_prefill.py:
```diff
@@ -0,0 +1,98 @@
+    print("Waiting for prefill node to finish...")
+    prefill_done.wait()
+
+    outputs = llm.generate(prompts, sampling_params)
```
```suggestion
    # At this point when the prefill_done is set, the kv-cache should have been transferred
    # to this decode node, so we can start decoding.
    outputs = llm.generate(prompts, sampling_params)
```

Shaoting-Feng replied:
Thanks for your comment. I've added this comment.

comaniac reviewed (commented):
Sorry for missing this PR. Overall LGTM. Please fix some more comments and we should be good to go.

Shaoting-Feng committed:
Add comments for better clarification

Signed-off-by: Shaoting <shaotingf@uchicago.edu>

Shaoting-Feng committed:
Add Apache-2.0

Signed-off-by: Shaoting <shaotingf@uchicago.edu>

comaniac reviewed (approved):
LGTM. Thanks!
